patroni/ha.py

```diff
@@ -175,12 +175,14 @@ def get_effective_tags(self):
 
     def touch_member(self):
         with self._member_state_lock:
+            primary_conninfo = self.state_handler.config._current_recovery_params.get('primary_conninfo', {})[0] if self.state_handler.config._current_recovery_params is not None and self.state_handler.role != 'master' else None
             data = {
                 'conn_url': self.state_handler.connection_string,
                 'api_url': self.patroni.api.connection_string,
                 'state': self.state_handler.state,
                 'role': self.state_handler.role,
-                'version': self.patroni.version
+                'version': self.patroni.version,
+                'master_host': primary_conninfo.get('host', None) if primary_conninfo is not None else None
```

Comment:
Thyamlere is no need to add a new field to the structure, [tags](https://github.com/zalando/patroni/pull/1576/files#diff-57c53e988b775ee2902488fb33c1f433R193-R195) already has everything you need, i.e.
```json
{
  "tags": {
    "replicatefrom": "foo"
  }
}
```

If there is a member `foo` in the cluster, the given member will be replicating from it (or at least try to do so).

Reply:
Yes, but tags are optional and the can be absent. So it this case you won't be able to build topology. Or I am not correct?

Reply:
It is not possible to configure cascading replication without using tags.

Reply:
Effectively the field you have added is just duplicating the `tags.replicatefrom`.
Absence of such tag means that the node is replicating from the leader.
When I told `it is trying to do so` I only meant to say that this node could be not in a good shape, for example postgres is failing to start. The node which holds the `replicatefrom` tag will stream from the primary only in one case, if the cascading node is not registered in DCS, therefore it is always safe to rely on tags.

If in the future there will be the other way of configuring cascading replication we will adjust the code which prints the topology accordingly, this is not a big deal, but now it is not necessary.

Reply:
I do not think it's duplicating. `replicatefrom` - is desired state, but this field is the current state.
If we will switch on using tags, we should also switch on using `role` field in dcs in order to determine who is current leader.

But let's assume the situation, where the current leader lost connection to dcs.
In this case there will be some time, when patroni will not report a leader, like 
```bash
+ Cluster: patroni14_test_mf_dba (6826270184262086410) --------+----+-----------+--------------------------------------------+
|            Member           |      Host     | Role |  State  | TL | Lag in MB | Tags                                       |
+-----------------------------+---------------+------+---------+----+-----------+--------------------------------------------+
| postgresql3 | 10.144.173.73 |      | running | 23 |       0.0 | replicatefrom: :undef                      |
+-----------------------------+---------------+------+---------+----+-----------+--------------------------------------------+
| postgresql4 | 10.144.173.74 |      | running | 23 |       0.0 | clonefrom: true                            |
|                             |               |      |         |    |           | replicatefrom: postgresql3 |
+-----------------------------+---------------+------+---------+----+-----------+--------------------------------------------+
```
but in actual topology, this cluster will look like 
```bash
Cluster: patroni14_test_mf_dba (6826270184262086410)
postgresql3  	|10.144.173.73:5432|replica|running|tl:23|
+ postgresql4	|10.144.173.74:5432|replica|running|tl:23|lag:0.0 MB|
```
so it's obviously has a master, although it might be in read-only mode.
and in case, if there was no cascade and tags are empty, we will see situation from the side of postgres like this (two standalone not replicated servers)
```bash
Cluster: patroni14_test_mf_dba (6826270184262086410)
postgresql3  	|10.144.173.73:5432|replica|running|tl:23|
postgresql4	|10.144.173.74:5432|replica|running|tl:23|lag:0.0 MB|
```

But In dcs they will both have `role:replica`.

if we will rely on `role` field in order to determine leader - it will make the logic of determining current master very complicated

Reply:
> so it's obviously has a master, although it might be in read-only mode.

Nope. The this topology example is incorrect. Absence of leader doesn't make any other node a new leader automatically., therefore there should be a `+ ` before postgresql3.

> if we will rely on role field in order to determine leader - it will make the logic of determining current master

That's the point. Only the node owning the leader key in the DCS is the true leader and you should never rely on the `role` field for that.

> I do not think it's duplicating. replicatefrom - is desired state, but this field is the current state.

This is sort of correct, but you can always derive one from another. If there are any minor discrepancies between these two states, they will go away in `loop_wait` seconds maximum.

After all, the member key in DCS is updated only once per HA loop, therefore it is not possible to avoid minor discrepancies.

Reply:
I think we are mixing point of views :)
You are speaking from the patroni side, how it sees the cluster, while I am from the postgresql side.
> Nope. The this topology example is incorrect. Absence of leader doesn't make any other node a new leader automatically., therefore there should be a `+ ` before postgresql3.

It's taken from the real cluster, where I've emulated connection issue between patroni on leader and dcs with large enough retry_timeout. While leader is retrying and postgresql is alive on it, during retry timeout there will be a situation, when leader key will be missing from dcs. So from the patroni side, this cluster do not have a leader. But, if we look into postgres, there obviously will be one master (`postgresql3`) and replica (`postgresql4`), as it's shown on topology. 

Moreover, the absence of leader key in such situations makes building topology graph very complicated, as there will be no vertex, from which we can start traversing it.

I understand that scenario like this with dcs failures are rare enough, but I think that presenting a true topology in such cases is very important in order to make decisions on real state of things.

Reply:
> You are speaking from the patroni side, how it sees the cluster, while I am from the postgresql side.

There is no point of view from the postgres side, the DCS is the only source of trust!

> It's taken from the real cluster, where I've emulated connection issue between patroni on leader and dcs with large enough retry_timeout.

So you violated the requirement to set `ttl <= loop_wait + 2 * retry_timeout `?
The master should be demoted in the worst case when the leader key expires, or better a bit earlier.

> So from the patroni side, this cluster do not have a leader. But, if we look into postgres, there obviously will be one master (postgresql3) and replica (postgresql4), as it's shown on topology.

Nope. The postgresql3 is the replica, and currently there is absolutely no guaranty that it will become the new leader.

> Moreover, the absence of leader key in such situations makes building topology graph very complicated, as there will be no vertex, from which we can start traversing it.

Just replace the `leader` with `None`. Problem solved.

Reply:
Yes, I understand your concerns and that this situation is a bit fictional and that is a misconfiguration, which I made for a purpose.
But let's think a bit further - as `replicatefrom` is user-configurable, there is always an opportunity of mistakes or typos. And the data I'm using for `master_host` is just the same methods as Patroni uses in order to fill `recovery.conf`. So if, for example, user will make a mistake in configuration, he will be able to see real topology and fix it

Reply:
> replicatefrom is user-configurable, there is always an opportunity of mistakes or typos. And the data I'm using for master_host is just the same methods as Patroni uses in order to fill recovery.conf. So if, for example, user will make a mistake in configuration, he will be able to see real topology and fix it.

In case of any typos or missing intermediate node it will be replicating from the leader:
https://github.com/zalando/patroni/blob/6406b39b77269252fc0b0149ca4472d0bab0f06c/patroni/ha.py#L366-L367
https://github.com/zalando/patroni/blob/6406b39b77269252fc0b0149ca4472d0bab0f06c/patroni/dcs/__init__.py#L445-L446

And I don't see any reason to not use the same logic in patronictl.

> may be it will be a good idea to add tags to output of topology, so such misconfigurations/typos be visible to user?

Probably, but currently tags are formatted to multiple lines, so drawing ASCII tree will become harder, unless we apply special formatting to tags (default_flow_style=True)

Reply:
I understand your concerns, that adding to structure in DCS a new field, that will be used only in one function can look like something wasteful, so will try to derive it from available info. 

But I got some question regarding finding who is a leader now.
https://github.com/zalando/patroni/blob/e95e54b94e08a9bdf624ba6cd126c26c6c95ff94/patroni/ctl.py#L765-L772

When we are using `list` command, we derive leader from member role. 
But you said that 

> That's the point. Only the node owning the leader key in the DCS is the true leader and you should never rely on the role field for that.

So what is the right way to find out who is a leader?

And one more question, why do patronictl set `role='' if m['role'] == 'replica'`?

Reply:
The "real" role is calculated here: https://github.com/zalando/patroni/blob/798c46bc0376f9b401bc25e45ef3be9745cc58f8/patroni/utils.py#L381-L393

And it guarantees that there is not more than one `leader` or `standby_leader` among members.

Reply:
> And one more question, why do patronictl set role='' if m['role'] == 'replica'?

Because it was always shown as an empty string: https://github.com/zalando/patroni/blob/dcb5113f9d937db498847cea2f0ad92b301d43cc/patroni/ctl.py#L328-L346

Reply:
Another reason to not introduce a new field - the new patronictl will work with old clusters.
It is quite common that people run patronictl remotely, not from the host where Patroin is running.

Reply:
> Because it was always shown as an empty string:
Can I fix it and return `replica` and not an empty string? I just want to reuse the same code, that is used in `output_members`?

Reply:
I don't have any objections on that.

Reply:
done, now use `replicatefrom` in order to find leader